src/controller.js

```diff
@@ -0,0 +1,601 @@
+    }
+
+    if (foundKey !== key) {
+      throw new Error('Document contains incorrect Key');
+    }
+
+    if (typeof lookingfor === 'undefined') {
+      throw new Error('Document does not contain ' + property);
+    }
+
+    return lookingfor;
+  }
+  /**
+   * Obtain an UploadId from an Initiate Multipart Upload response body
+   */
+  __getUploadId(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'InitiateMultipartUploadResult', 'UploadId', bucket, key);
+  }
+
+  /**
+   * Obtain the ETag for the committed multipart upload
+   */
+  __getMultipartEtag(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'CompleteMultipartUploadResult', 'ETag', bucket, key);
+  }
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
+    permissions = runSchema(permissions, schemas.permissions.required());
+    let {acl, read, write, readAcp, writeAcp, fullControl} = permissions;
+    if (acl) {
+      return [['x-amz-acl', permissions.acl]];
+    }
+
+    let perms = [];
+
+    if (read) {
+      perms.push(['x-amz-grant-read', read]);
+    }
+
+    if (write) {
+      perms.push(['x-amz-grant-write', write]);
+    }
+
+    if (readAcp) {
+      perms.push(['x-amz-grant-read-acp', readAcp]);
+    }
+
+    if (writeAcp) {
+      perms.push(['x-amz-grant-write-acp', writeAcp]);
+    }
+
+    if (fullControl) {
+      perms.push(['x-amz-grant-full-control', fullControl]);
+    }
+
+    return perms;
+  }
+
+  __s3hostname(bucket) {
+    let hostname = `${bucket}.${this.s3host}`;
+    if (this.s3port) {
+      hostname += ':' + this.s3port;
+    }
+    return hostname;
+  }
+ 
+  /**
+   * Initiate a Multipart upload and return the UploadIp that
+   * Amazon has assigned for this multipart upload
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html
+   */
+  async initiateMultipartUpload(opts) {
+
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.mpSize.required(),
+      permissions: schemas.permissions,
+    }).optionalKeys('permissions'));
+
+    let {bucket, key, sha256, size, permissions} = opts;
+
+    if (size <= 0) {
+      // Each part must have a non-zero size
+      throw new Error('Objects must be more than 0 bytes');
+    } else if (size > 5 * 1024 * 1024 * 1024 * 1024) {
+      // The entire file must be lower than 5 TB
+      throw new Error('Object must total fewer than 5 TB'); 
+    }
+    let unsignedRequest = {
+      service: 's3',
+      region: this.region,
+      method: 'POST',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploads=`,
+      headers: {
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
```

Comment:
except that we sort of have to do it.

Content-Length is part of V4 signing when specified, so we need to pass the actual content-length for each call to AWS.  For single part uploads, that's no problem because the content-length is the same as the artifact size.  For multipart uploads, however, we need to set the content-length header to the artifact size, but that'd cause the signing of the initiate multipart upload to fail because that request would have the wrong content-length.

S3 might calculate a valid Content-Length at the end of the multipart upload, but it is not a value directly passed through by the worker.  That's why we need to have `x-amz-meta-content-length`: it's the exact length the worker thought the file is.

I set this for single part uploads as well so that they have the same structure.

I'll add support for content-type.